src/os/bluestore/BlueStoreSlowFastCoDel.h

```diff
@@ -0,0 +1,152 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+
+#pragma once
+
+#include <iostream>
+#include <atomic>
+
+#include "include/Context.h"
+#include "include/utime.h"
+#include "common/Thread.h"
+#include "common/ceph_time.h"
+
+class BlueStoreSlowFastCoDel {
+public:
+  BlueStoreSlowFastCoDel(
+    CephContext *_cct,
+    std::function<void(int64_t)> _bluestore_budget_reset_callback,
+    std::function<int64_t()> _get_kv_throttle_current);
+
+  virtual ~BlueStoreSlowFastCoDel();
+
+  void on_config_changed(CephContext *cct);
+
+  void reset_bluestore_budget();
+
+  void update_from_txc_info(
+    ceph::mono_clock::time_point txc_start_time,
+    uint64_t txc_bytes);
+
+  int64_t get_bluestore_budget();
+
+  int64_t get_target_latency();
+
+  bool is_activated();
+
+protected:
+  struct CoDelLoopThread : public Thread {
```

Comment:
I assume that the goal here is to avoid the assert failure you were seeing before.  The purpose of that assert is to enforce an invariant in SafeTimer that calling add_event_after needs to be atomic with cancellation and with actually executing the callbacks.

Your original implementation uses register_lock to enforce atomicity between codel cycles (*_interval_process).  You also chose to create two new locks

```
  ceph::mutex fast_timer_lock = ceph::make_mutex("CoDel::fast_timer_lock");
  ceph::mutex slow_timer_lock = ceph::make_mutex("CoDel::slow_timer_lock");
```

to use with the SafeTimer instances:

```
  fast_timer(_cct, fast_timer_lock),
  slow_timer(_cct, slow_timer_lock),
```

This means that _fast_interval_process and _slow_interval_process are called by SafeTimer while holding fast_timer_lock and slow_timer_lock respectively, so in normal operation the calls to add_next_event aren't a problem.